kubernetes/updater/pkg/controller/updater.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+
+	"github.com/docker/distribution/reference"
+	"github.com/gravitational/trace"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/gravitational/teleport/kubernetes/updater/pkg/img"
+	"github.com/gravitational/teleport/kubernetes/updater/pkg/maintenance"
+	"github.com/gravitational/teleport/kubernetes/updater/pkg/version"
+)
+
+type VersionUpdater struct {
+	versionGetter       version.Getter
+	imageValidators     []img.Validator
+	maintenanceTriggers []maintenance.Trigger
+	baseImage           reference.Named
+}
+
+// TODO: find a better name
+func (r *VersionUpdater) Do(ctx context.Context, obj v1.Object, currentVersion string) (img.NamedTaggedDigested, error) {
+	// Those are debug logs only
+	log := ctrllog.FromContext(ctx).V(1)
+
+	// Can we do a maintenance?
+	log.Info("Checking if a maintenance trigger is on")
+	if !maintenance.AtLeastOne(ctx, obj, r.maintenanceTriggers...) {
+		return nil, &MaintenanceNotTriggeredError{}
+	}
+	log.Info("Maintenance triggered, getting new version")
+
+	// Get the next version
+	nextVersion, err := r.versionGetter.GetVersion()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	log.Info("New version candidate", "nextVersion", nextVersion)
+	if !version.ValidVersionChange(currentVersion, nextVersion, "") {
+		return nil, &NoNewVersionError{CurrentVersion: currentVersion, NextVersion: nextVersion}
+	}
+
+	log.Info("Version change is valid, building img candidate")
+	// We tag our img candidate with the version
+	image, err := reference.WithTag(r.baseImage, nextVersion[1:])
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	log.Info("Verifying candidate img", "img", image.String())
+	// We validate the signatures
+	valid, digested := img.AtLeastOneValid(ctx, image, r.imageValidators...)
```

Comment:
Can we change this to be a function of the set of image validators?

```go
type Validators []img.Validators

func (v Validators) AtLeastOneValid(ctx, image){}


```

It feels easier to read

Reply:
Also done for `maintenance.Triggers`